### ISSUE TITLE
Disconnect/Reconnect every 2-5 seconds

### ISSUE THREAD

**Issue by jswigart** [Celtoys/Remotery#43]
There are a couple calls to Server_Send with a 20ms timeout that frequently appear to fail inside TCPSocket_Send in the timeout busy loop, forcing a disconnect. This happens every few seconds. I'm not sure why it would be timing out on a loopback connection.

If it matters, I'm running the web page in the latest chrome.

**Comment by dwilliamson** [comment]
Can you point me to the exact line of failure?

Are you sending a lot of sample data each frame?

What is your value for `rmt->json_buf->bytes_used` at https://github.com/Celtoys/Remotery/blob/master/lib/Remotery.c#L3851 ?

**Comment by jswigart** [comment]
Yes I was sending several hundred k at a time, partly because of the profiling in a loop that doesn't merge samples as I mentioned in another issue causing a ton of data bloat. I removed those inner samples for the time being and am not seeing this anymore so far.

**Comment by dwilliamson** [comment]
OK, that's good. Although I'd like to find some way of gracefully rejecting large data without disconnecting all the time so I'll leave this one open.

**Comment by hlide** [comment]
It looks like I have this issue as well. I use NvPA stuff () but I just saw your stuff and found it interesting but it looks like Remotery has issue with number of events or maybe with very short events (some can be < 1ms) because I fail to make Remotery in the same place where I have NvPA working.

**Comment by alpha-xue** [comment]
Hi @hlide @hjanetzek, I am a student in university, and I am planning to show one CUDA profiling in web broswer. I am not sure how to take use of these code, could you please give some guide? My Lab setup as following: 1)I have PHP&Apache server install in one linux as webserver. 2) I have 1 GPU server and CUDA programs running on this GPU server. 3)I want to show profiling procedure and result for my CUDA program in remote Firefox or Chrom(e.g. my windows laptop). Is it possible to take use these codes? How to setup the server address and compile them? 

I am really appreciate your input here as I am stuck with my assignment here.

**Comment by alpha-xue** [comment]
Hi @hlide @hjanetzek @jswigart , I am a student in university, and I am planning to show one CUDA profiling in web broswer. I am not sure how to take use of these code, could you please give some guide? My Lab setup as following: 1)I have PHP&Apache server install in one linux as webserver. 2) I have 1 GPU server and CUDA programs running on this Linux GPU server. 3)I want to show profiling procedure and result for my CUDA program in remote Firefox or Chrom(e.g. my windows laptop). Is it possible to take use these codes? How to setup the server address and compile them? 

I am really appreciate your input here as I am stuck with my assignment here.

**Comment by dwilliamson** [comment]
@alpha-xue please refer to the readme where it shows you how to run the viewer and instrument your code

https://github.com/Celtoys/Remotery#basic-use

The viewer will initially open a connection to `ws://127.0.0.1:17815/rmt`. You need to change the IP address to the computer you wish to profile and ensure the port has access. Nothing more is required.

**Comment by dwilliamson** [comment]
@alpha-xue I'll remove your comment above in a few days to give you a chance to read it. I want to keep issue threads as clear as possible.

If you have an issue with Remotery not working or a feature request, please open a new issue and I can deal with it. This is not a general C support forum, however - you might find a lot more help at websites such as http://gamedev.net

**Comment by alpha-xue** [comment]
@dwilliamson Thanks for you replay. I read the whole introduction carefully but I still have following question and concern:
1. Even I change url from  ws://127.0.0.1:17815/rmt  mine specific, my question is where can we define or change the server port:17851. Can I use different port let server side listening different port?
2. I download the zip source file and try to compile the original source code but it failed serveral time. I resolve them one but one but still fail.
Can I send email for your check? really appreciate you can do me one favor on this issue as I am under greate course pressure. Thanks a lot. Mine email address alpha.xue.future@gmail.com

**Comment by alpha-xue** [comment]
I am not sure if you can receive this email but I really need you expert
help about this issue.

Thanks and regards!

2016-02-20 0:48 GMT+08:00 Don Williamson <notifications@github.com>:

**Comment by zao** [comment]
Dear @alpha-xue,

Everyone sees the replies you make to an "issue". Any email address you see leads back to the Github issues, where replies go into the corresponding issue thread.
If you have a new problem with the codebase, please open a new issue to avoid polluting this current one, as everyone who watches it gets the messages.

As for being under pressure, we all are. If you have trouble adapting the project for your needs, reconsider if you need the project or consult with your teaching assistants or professors.

**Comment by alpha-xue** [comment]
@zao Thanks, I know it is not suitable to replay or comment agian and again here. If you have time, could you spend half an hour on this issue? Really appreciate it.

**Comment by dwilliamson** [comment]
@zao thankyou for your comment, deleting all now

**Comment by alpha-xue** [comment]
Thanks Don.

2016-02-21 8:47 GMT+08:00 Don Williamson <notifications@github.com>:

**Comment by samswain** [comment]
I've hit this too.  Seems to happen when my app frame rate goes pretty choppy.  Ironic as this is when I need to know what's going on.  :)

**Comment by dwilliamson** [comment]
Are you sending lots of sample points each frame?

**Comment by samswain** [comment]
Not really.  I have several threads doing various things at various rates, some are long task ones, others (mainly the renderer) are regular.  I have some stalls though that may interrupt things a bit.  Problem seems far worse in debug/dev builds (i.e. slower).  Will do some more investigation and let you know.

**Comment by Celtoys** [comment]
If you've got lots of stalls you may somehow be starving the browser process and flooding the socket connection. I've never tested that kind of scenario before.

**Comment by dwilliamson** [comment]
Just noticed that if you have threads that are starving the OS network stack (e.g. you're busy-waiting on a spin-lock for too long, or something) then any timeout during a sample tree send has to be considered fatal.

This is because it's done in [two steps, rather than one](https://github.com/Celtoys/Remotery/blob/master/lib/Remotery.c#L2791), to reduce the number of runtime memory allocations required.

You can detect that this is the case if you [breakpoint this line](https://github.com/Celtoys/Remotery/blob/master/lib/Remotery.c#L1982).

Would take a little rework to embed the Websocket header before each message instead of using two sends (and **without** allocating a new buffer) but that should solve any problems.

**Comment by Celtoys** [comment]
Should be fixed with 5ed038ead9562e603581de92665def34899e8beb
